Refactor `create_geometry` and add support for retina half axes and nonzero pupil thickness

crnh opened the pull request:
- Specify parameters for `create_geometry` as a `TypedDict`
- Add new `retina_axial_half_axis` and `retina_radial_half_axis` parameters to configure the retina using its half axes instead of its radius and asphericity. A `ValueError` is raised if both a radius / asphericity and a half axis are specified, or if only a single half axis is specified.
- Add a new `pupil_lens_distance` parameter to `create_geometry` and a property with the same name to `EyeGeometry`, to allow for a nonzero distance between the pupil and the lens.

@jwmbeenakker do you agree with the names of the new parameters, or do you prefer different names?

jwmbeenakker reviewed (changes requested):
To me it is not entirely clear what you mean with `half_axis` is it the radius/semi-diameter or something else?

crnh commented:
> To me it is not entirely clear what you mean with `half_axis` is it the radius/semi-diameter or something else?

It's 1/2 of the ellipse axis (mathematically we're usually speaking about the semi-major and semi-minor axes, but these terms don't work in our case because they don't convey information about the directionality).
I prefer to avoid terms like 'radius' and 'semi-diameter' for these dimensions, because 'radius' is already used for the radius of curvature (which is related to the half-axes, but not equal), and semi-diameter is already used for the aperture. Using any of these terms here will likely lead to more confusion. 
Semi-axis could work here, but I am not sure if that makes it much clearer.

jwmbeenakker commented:
> > To me it is not entirely clear what you mean with `half_axis` is it the radius/semi-diameter or something else?
> 
> It's 1/2 of the ellipse axis (mathematically we're usually speaking about the semi-major and semi-minor axes, but these terms don't work in our case because they don't convey information about the directionality). I prefer to avoid terms like 'radius' and 'semi-diameter' for these dimensions, because 'radius' is already used for the radius of curvature (which is related to the half-axes, but not equal), and semi-diameter is already used for the aperture. Using any of these terms here will likely lead to more confusion. Semi-axis could work here, but I am not sure if that makes it much clearer.

hmm, indeed tricky. for me therm axis is more related a direction (like optical-axis, x-axis) and not to a dimension. I agree with not wanting to use `radius` as it is too close to radius-of-curvature, which is often used in the same context. I would personally go for `semi-diameter`, since the ellipsoid-surfaces are not apertures, so the context makes it clear

crnh commented:
>  I would personally go for semi-diameter, since the ellipsoid-surfaces are not apertures, so the context makes it clear

True, but they are still optical surfaces with a `semi_diameter` property... So I am not sure if naming a property `retina_radial_semi_diameter` if it does not set the `semi_diameter` of the surface will reduce the confusion. 
The dimensions are actually related to the axes you mentioned, as the axial half-axis is parallel to the optical axis and the radial half-axis parallel to the x/y-axis.

crnh commented:
I renamed the `half_axes` to `ellipsoid_radii` and also renamed similar properties on the `StandardSurface` class. Also, I renamed `axial` to `z` and `radial` to `y`, to allow for non-rotationally symmetric retinas in the future. 
`StandardSurface.ellipsoid_radii` returns a tuple of radii along the z, y and x axes. If the surface is rotationally symmetric, `x` is set to the value of the `y` radius.
